feat: flashcard generation feature

Wenle611 opened the pull request:
This pull request implements the complete flashcard generation feature (Use Case 3), enabling users to generate study flashcards from course materials using the Google Gemini AI API.

**Flashcard Generation Core:**

- Added `GeminiFlashcardGenerator`, a new data access class that integrates with the Gemini AI API to generate flashcards from PDF files using configurable prompts and JSON schemas.
- Implemented `FlashcardGenerator` interface following Dependency Inversion Principle for easy testing and future extensibility.
- Added `HardCodedCourseLookup` as a temporary data access implementation (to be replaced with proper CourseRepository integration).

**Clean Architecture Components:**

- Added `GenerateFlashcardsInteractor` as the use case logic with comprehensive error handling for both IOException and RuntimeException.
- Added `GenerateFlashcardsController` to handle user input from the UI layer.
- Added `GenerateFlashcardsPresenter` to update ViewModel and manage view navigation via ViewManagerModel.
- Added `FlashcardViewModel` for UI state management using PropertyChangeSupport pattern.

**User Interface:**

- Added `GenerateFlashcardsView` - Input form for course name and topic selection.
- Added `FlashcardDisplayView` - Interactive flashcard study interface with:
  - Flip animation to toggle between question (front) and answer (back)
  - Navigation controls (Previous/Next buttons)
  - Progress tracking (Card X / Y)
  - Back button to return to Course Workspace

**Integration:**

- Enhanced `AppBuilder` class to initialize and wire up flashcard views, view models, and controllers with proper dependency injection.
- Connected "Create Flashcards" button in `CourseWorkspaceView` to trigger flashcard generation.
- Implemented view navigation using `ViewManagerModel` for seamless transitions.

**Entities:**

- Added `Flashcard` entity representing a single question-answer pair.
- Added `FlashcardSet` entity representing a collection of flashcards for a course.

**Testing:**

- Added unit tests for `GenerateFlashcardsInteractor` covering:
  - Successful flashcard generation
  - IOException handling
  - RuntimeException handling

**Related User Story:**
- As a user, I want to have flash cards automatically generated based on my uploaded course materials.

Wenle611 committed:
Implement Flashcard Generation use case and FlashcardGenerator

Wenle611 committed:
Implement Gemini PDF flashcard generation

yupeiyu4 commented:
I noticed the PR includes local machine artifacts (e.g., .DS_Store files), which should be removed from version control and ignored so they don’t keep reappearing in future commits. I also recommend adding unit tests for the core interactor to cover the main success path and the key failure paths (e.g., invalid inputs and exceptions from the generator), and ideally including a coverage report screenshot to demonstrate strong test coverage for the use-case logic. Finally, please double-check Clean Architecture boundaries: keep UI code out of the use case layer, depend on gateway interfaces rather than concrete data-access classes, keep naming/formatting consistent with the project’s style, and add brief Javadocs for any new public classes/methods; a short “flow of control” note in the PR description would also make the implementation easier to review.

shirleyy-z commented on src/main/java/entities/Flashcard.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    // Overloaded constructor for explicitly defining front/back
+    public Flashcard(String front, String back, boolean isFrontBack) {
```
The isFrontBack parameter isn’t used anywhere, which is a bit confusing. If the semantics are always “front/back == question/answer”, can we drop the boolean, or use it to distinguish behaviour in some way? Not sure

shirleyy-z commented on src/main/java/app/Main.java:
```diff
@@ -1,24 +1,8 @@
-        application.setLocationRelativeTo(null); // Center the window
-        application.setVisible(true);
+        AppBuilder app = new AppBuilder();
+        app.runFlashcardGeneration();
```
This switches the entry point from the Swing UI to a console-style flashcard demo, which is fine for now. Can we add a short comment indicating this is a temporary demo path so teammates/TA know why the GUI is no longer launched?

shirleyy-z commented on src/main/java/app/AppBuilder.java:
```diff
@@ -1,102 +1,53 @@
+    public void runFlashcardGeneration() {
+        try {
+            HardCodedCourseLookup lookup = new HardCodedCourseLookup();
+            Course course = lookup.getCourseById("RLG200");
```
Hard-coding the course ID means this demo only works for RLG200. Maybe make the course ID a parameter or configuration so we can reuse this flow for other courses?

Wenle611 replied:
I'll parameterize this once I integrate with the actual UI. Keeping it simple for the demo for now.

shirleyy-z commented on src/main/java/app/AppBuilder.java:
This class used to wire up the Swing UI, but now it mainly runs a flashcard generation demo. Maybe it would make sense to move the demo into a separate class (e.g., FlashcardDemoRunner) and keep AppBuilder focused on building the application / GUI wiring?

shirleyy-z commented on src/main/java/entities/FlashcardSet.java:
This is a nice encapsulation of a set of flashcards for a course.

shirleyy-z reviewed (commented):
From a Clean Architecture / SOLID perspective, the new use case and data-access pieces are quite well-structured. Defining FlashcardGenerator as an interface and injecting it into GenerateFlashcardsInteractor follows the Dependency Inversion Principle and keeps the core logic decoupled from the concrete Gemini client. Similarly, the use of input/output boundaries and a response model is consistent with the “use cases in the middle” pattern and will make it easier to swap the console presenter for a GUI presenter later. The new entities are small and focused, and the GeminiFlashcardGenerator cleanly encapsulates the external API interaction.

Architecturally, the main concern is around AppBuilder and Main. AppBuilder previously built the Swing application; in this PR it’s been partially converted into a one-off flashcard demo runner while still retaining some remnants of the old GUI wiring. The new HardCodedCourseLookup is fine as a temporary data-access implementation, but longer-term it should be hidden behind an interface so higher layers depend on abstractions rather than a specific “demo” concrete implementation.

Also, be sure to add in unit testing for your GernateFlashcardsInteractor! Make sure you can reach 100% code coverage.

In terms of style and correctness, the new code is very readable and follows consistent naming.

Overall, very well done!

Wenle611 commented on src/main/java/data_access/HardCodedCourseLookup.java:
```diff
@@ -0,0 +1,24 @@
+import entities.PDFFile;
+
+public class HardCodedCourseLookup {
+
```
Good point! I'll add a comment noting this is temporary and will be replaced by a proper course repository.

Wenle611 committed:
Merge branch 'main' into feat/flashcards-generation

Wenle611 committed:
Remove unused isFrontBack constructor from Flashcard entity

Wenle611 committed:
Add documentation noting HardCodedCourseLookup is temporary

Wenle611 committed:
Add FlashcardViewModel for UI state management

Wenle611 committed:
Add GenerateFlashcardsController to handle user input

Wenle611 committed:
Update GenerateFlashcardsPresenter to work with ViewModel

Wenle611 committed:
WIP: Add flashcard views (incomplete)

Wenle611 committed:
Complete flashcard feature with navigation

Wenle611 committed:
feat: integrate flashcard generation feature

Wenle611 committed:
fix: complete flashcard feature with proper navigation and error handling

- Fix false error dialog by checking null before showing error message
- Add ViewManagerModel to GenerateFlashcardsPresenter for view navigation
- Implement back button navigation in FlashcardDisplayView
- Wire up flashcard views in AppBuilder with proper dependencies
- Flashcard generation now correctly displays results after API call

Wenle611 committed:
test: add unit tests for GenerateFlashcardsInteractor

- Test successful flashcard generation
- Test IOException handling
- Test RuntimeException handling
- Add RuntimeException catch block in Interactor

enxilium commented on src/main/java/views/CourseWorkspaceView.java:
This is test code, it can be removed before merging to main.

enxilium commented on src/.DS_Store:
Unneeded file for production

enxilium reviewed (commented):
Overall, excellent use of clean architecture! Just a few unnecessary files were accidentally committed, and a few comments need to be cleaned up before we can merge to main. Great job!

Wenle611 committed:
fix: repair corrupted .gitignore file

Wenle611 committed:
chore: remove .idea from version control, fix pom.xml

Wenle611 commented on src/main/java/app/AppBuilder.java:
```diff
@@ -192,6 +200,52 @@ public AppBuilder addCourseUseCases() {
         return this;
     }
 
+    // === FLASHCARD: Flashcard methods ===
```
Thanks!

Wenle611 committed:
refactor: consolidate Gemini API and clean up debug code

- Add FlashcardGenerator implementation to GeminiApiDataAccess
- Remove separate GeminiFlashcardGenerator class
- Remove debug print statements from CourseWorkspaceView
- Update AppBuilder to use shared geminiDAO

enxilium commented:
Closing this PR as I've merged the code in the CLI.
